// nodeblog-ext-dynamic-router/index.js
"use strict";

var path = require('path');
var router = require(path.join(LIBS_DIR, 'router.js'))();
var emitter = require(path.join(LIBS_DIR, 'eventEmitter'));

module.exports = function() {
    var self = {};

    self.registerStaticRoutes = function(nb) {
        emitter.emit('ext:dynamic-router:beforeRegisterStaticRoutes', nb);

        /***** set admin route *****/
        var adminRouter = require('./admin');
        nb.app.use('/admin', adminRouter);

        router.addRoute('login', {
            path: '/login',
            get: 'user:loginForm',
            post: 'user:checkLogin',
            listing: false,
            title: 'User login'
        });

        router.addRoute('logout', {
            path: '/logout',
            get: 'user:logout',
            listing: false,
            title: 'User logout'
        });

        router.addRoute('Page', {
            path: '/:page_url.html',
            title: 'Page',
            get: 'page:view',
            params_handle: {
                page_url: function (req, res, next, page_url) {
                    res.end(page_url);
                }
            }
        });
    };

    return self;
};
